specs/language/expressions.tex

```diff
@@ -160,3 +160,86 @@
 overloaded implementation of \texttt{operator[]} (\ref{Overload}).\footnote{HLSL
 does not support the base address of a subscript operator being the expression
 inside the braces, which is valid in C and C++.}
+
+\Sec{Function Calls}{Expr.Post.Call}
+
+\p A function call may be an \textit{ordinary function}, or a \textit{member
+function}. In a function call to an \textit{ordinary function}, the
+\textit{postfix-expression} must be an lvalue that refers to a function. In a
+function call to a \textit{member function}, the \textit{postfix-expression}
+will be an implicit or explicit class member access whose \textit{id-expression}
+is a member function name.
+
+\p When a function is called, each parameter shall be initialized with its
+corresponding argument. The order in which parameters are initialized is
+undefined. \footnote{Today in DXC targeting DXIL matches the Microsoft C++ ABI
+and evaluates argument expressions right-to-left, while SPIR-V generation
+matches the Itanium ABI evaluating parameters left-to-right. There are good
+arguments for unifying these behaviors, and arguments for keeping them
+different.}
+
+\p If the function is a non-static member function the \texttt{this} argument
+shall be initialized to a reference to the object of the call as if casted by an
+explicit cast expression.
+
+\p Parameters are either \textit{input parameters}, \textit{output parameters},
+or \textit{input/output parameters} as denoted in the called function's
+declaration (\ref{Decl.Function}).
+
+\p \textit{Input parameters} are passed by-value into a function. If an argument
+to an \textit{input parameter} is of constant-sized array type, the array is
+copied to a temporary and the temporary value is converted to an address via
+array-to-pointer decay. If an argument is an unsized array type, the array
+lvalue directly decays via array-to-pointer decay. \footnote{This results in
+\textit{input} parameters of unsized arrays being modifiable by a function.}
+
+\p Arguments to \textit{output} and \textit{input/output parameters} must be
+lvalues. \textit{Output parameters} are not initialized prior to the call; they
+are passed as an uninitialized cx-value (\ref{Basic.lval}). An \textit{output
+parameter} is only initialized explicitly inside the called function. It is
+undefined behavior to not explicitly initialize an \textit{output parameter}
+before returning from the function in which it is defined. The cx-value created
+from an argument to an \textit{input/output parameter} is initialized through
+copy-initialization from the lvalue argument expression. In both cases, the
+cx-value shall have the type of the parameter and the argument can be converted
+to that type through implicit or explicit conversion.
```

Comment:
> the argument can be converted to that type through implicit or explicit conversion

Is this saying that the implementation has a choice of how to do it?  What's the difference between implicit or explicit conversion here?  (At this point, isn't everything implicit?)

Reply:
A user can explicitly convert argument types. Some casts in HLSL are allowed to produce lvalues (which is a bit odd). An example of this in practice is:

```c++
struct Doggo {
  int Legs;
  bool Tail;
};

struct Kitten {
  int Legs;
  bool Teeth;
};

void Pet(inout Doggo D) {}

void main() {
  Kitten K = {4, true};
  Pet((Doggo)K);
}
```
[Compiler Explorer](https://godbolt.org/z/5jM63voa1)

In HLSL 2018, the case above was allowed implicitly or explicitly, in HLSL 2021+ it is only allowed explicitly. An example that HLSL 2021 allows implicitly would be:

```c++
void Trunc(int3 I) {}

void main() {
  float3 V = {1.5, 2.5, 3.5};
  Trunc(V);
}
```
[Compiler Explorer](https://godbolt.org/z/7o5e5hWzT)

Reply:
aside: we should probably implement a warning for stuff like this use of `Trunc`... I doubt anyone expects it to work how it does.

Reply:
Agreed. DXC has some warnings for implicit conversions, but not nearly enough.